diff --git a/fs/proc/base.c b/fs/proc/base.c
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -97,6 +97,10 @@
 #include "fd.h"
 #include "internal.h"

+#ifdef CONFIG_KSU_SUSFS_SUS_PROC
+#include <linux/susfs.h>
+#endif
+
 #include "../../lib/kstrtox.h"

 /* NOTE:
@@ -3398,8 +3402,19 @@ struct dentry *proc_pid_lookup(struct dentry *dentry, unsigned int flags)
 	if (!task)
 		goto out;

+#ifdef CONFIG_KSU_SUSFS_SUS_PROC
+	{
+		char comm[TASK_COMM_LEN];
+		get_task_comm(comm, task);
+		if (susfs_should_hide_proc(comm)) {
+			put_task_struct(task);
+			goto out;
+		}
+	}
+#endif
+
 	result = proc_pid_instantiate(dentry, task, NULL);
 	put_task_struct(task);
 out:
 	return result;
 }
@@ -3702,6 +3717,17 @@ int proc_pid_readdir(struct file *file, struct dir_context *ctx)
 		char name[10 + 1];
 		unsigned int len;

+#ifdef CONFIG_KSU_SUSFS_SUS_PROC
+		if (iter.task) {
+			char comm[TASK_COMM_LEN];
+			get_task_comm(comm, iter.task);
+			if (susfs_should_hide_proc(comm)) {
+				iter = next_tgid(ns, iter);
+				continue;
+			}
+		}
+#endif
+
 		cond_resched();
 		if (!has_pid_permissions(file, iter.task, HIDEPID_INVISIBLE))
 			goto out;
