diff --git a/fs/open.c b/fs/open.c
--- a/fs/open.c
+++ b/fs/open.c
@@ -35,6 +35,10 @@

 #include "internal.h"

+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+#include <linux/susfs.h>
+#endif
+
 int do_truncate(struct user_namespace *mnt_userns, struct dentry *dentry,
 		loff_t length, unsigned int time_attrs, struct file *filp)
 {
@@ -396,6 +400,13 @@ static int do_faccessat(int dfd, const char __user *filename, int mode, int flag
 	struct vfsmount *mnt;
 	int res;

+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+	if (susfs_check_unicode_bypass(filename)) {
+		return -ENOENT;
+	}
+#endif
+
 	if (mode & ~S_IRWXO)	/* where's F_OK, X_OK, W_OK, R_OK? */
 		return -EINVAL;

@@ -1196,6 +1207,13 @@ static long do_sys_openat2(int dfd, const char __user *filename,
 	if (unlikely(open_flags != op->open_flag))
 		return -EINVAL;

+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+	if (susfs_check_unicode_bypass(filename)) {
+		return -ENOENT;
+	}
+#endif
+
 	fd = build_open_flags(op, &flags);
 	if (fd)
 		return fd;
diff --git a/fs/stat.c b/fs/stat.c
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -20,6 +20,10 @@
 #include <linux/uaccess.h>
 #include <asm/unistd.h>

+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+#include <linux/susfs.h>
+#endif
+
 #include "internal.h"
 #include "mount.h"

@@ -217,6 +221,13 @@ int vfs_fstatat(int dfd, const char __user *filename,
 {
 	int ret;

+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+	if (filename && susfs_check_unicode_bypass(filename)) {
+		return -ENOENT;
+	}
+#endif
+
 	if (flag & ~(AT_SYMLINK_NOFOLLOW | AT_NO_AUTOMOUNT | AT_EMPTY_PATH))
 		return -EINVAL;
